### DISCUSSION TITLE
How can I forward my sl4j logs which are generated in my spring microservices to otel-collector ?

### DISCUSSION THREAD

**Discussion by yatamraghu7** [open-telemetry/opentelemetry-java-instrumentation#11872] [Q&A] [answered]
Hi Fellow Developer,
In my spring microservices I am using sleuth with otel combination to forward my traces to otel-collector and I am successful in forwarding my traces. But not able to achieve forwarding my logs to otel-collector.
I can see that opentelemetry is supporting multiple logging mechanisms which are being used in java but I have not found anything related to sl4j. Can someone guide me how can i forward my sl4j logs to otel-collector.
Thanks in advance.
Reagrd,
Raghu

**Comment by laurit** [comment]
sl4j is a logging facade, actual logging is performed by an logging framework such as logback, log4j etc. Check what is the actual logging framework that you use.

**Reply by yatamraghu7** [reply]
Hi @laurit 
We are only using slf4j for logging purpose in our project. Can you please help me in understanding how my logs can be forwarded to otel-collector ?
Thanks

**Reply by laurit** [reply]
to reiterate, slf4j is a logging facade that delegates logging to another framework. See https://www.slf4j.org/manual.html#swapping

**Reply by yatamraghu7** [reply]
Hi @laurit ,
I have changed the logging framework to log4j and using the maven dependency as below with xml configuration as follows

	<dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-log4j-appender-2.17</artifactId>
			<version>2.6.0-alpha</version>
	</dependency>

XML Config

```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="WARN">
	<Appenders>
		<Console name="Console" target="SYSTEM_OUT">
			<PatternLayout
				pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} trace_id: %X{trace_id} span_id: %X{span_id} trace_flags: %X{trace_flags} - %msg%n" />
		</Console>
		<Appenders>
			<OpenTelemetry name="OpenTelemetryAppender"
				captureMapMessageAttributes="true" 
				captureMarkerAttribute="true"
				captureContextDataAttributes="*" />
		</Appenders>
	</Appenders>
	<Loggers>
		<Root>
			<AppenderRef ref="OpenTelemetryAppender" level="All" />
			<AppenderRef ref="Console" level="All" />
		</Root>
	</Loggers>
</Configuration>
```


With these settings i have added the configuration file in my java with below bean

 @Bean
    public SdkLoggerProvider logProvider() {
        SdkLoggerProvider loggerProvider = SdkLoggerProvider.builder()
                .addLogRecordProcessor(BatchLogRecordProcessor
                        .builder(OtlpHttpLogRecordExporter.builder().setEndpoint("http://localhost:4318").build())
                        .build())
                .build();
        return loggerProvider;
    }

With all these settings I can only see the metrics related to span. I would like to see the logs related to my request and responses can you please guide me if I have done anything wrong in above procedure

2024-07-23 16:05:20.623 UTC
span_metrics_duration_sum
2024-07-23 16:05:20.623 UTC
span_metrics_duration_sum
2024-07-23 16:05:20.623 UTC
span_metrics_duration_count
2024-07-23 16:05:20.623 UTC
span_metrics_duration_count
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket
2024-07-23 16:05:20.623 UTC
span_metrics_duration_bucket

Thanks
Raghu

**Reply by laurit** [reply]
Have a look at https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/log4j/log4j-appender-2.17/library did you include the call to `OpenTelemetryAppender.install(...)`. It should be easier to get started with either the [java agent](https://opentelemetry.io/docs/zero-code/java/agent/) or [spring boot starter](https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/) Both of these configure this automatically.

**Reply by yatamraghu7** [reply]
Hi @laurit 

I have included the openTelemetryAppender as well and I could not find the logs in my qryn backend which otel-collector is collecting. Any idea why i can only see the metrics but not the original logs itself. Also If i try to use the OtlpHttpLogExporter instead of OtlpHttpLogRecordExporter I am getting error that class was not found. I checked in the google form and can see that this OtlpHttpLogExporter is being used by some projects to export the logs.
Is it because I am using OtlpHttpLogRecordExporter instead of OtlpHttpLogExporter. I tried to use OtlpHttpLogExporter as well but I could not find which mvn dependency can provide this class.

```
@Bean
    LogRecordProcessor otelLogRecordProcessor() {
        return BatchLogRecordProcessor
                .builder(OtlpHttpLogRecordExporter.builder().setEndpoint("http://localhost:4318").build())
                .build();
    }
```

 ```
@Bean
    OpenTelemetry openTelemetry(SdkLoggerProvider sdkLoggerProvider, SdkTracerProvider sdkTracerProvider,
            ContextPropagators contextPropagators) {
        OpenTelemetrySdk openTelemetrySdk = OpenTelemetrySdk.builder()
                .setLoggerProvider(sdkLoggerProvider)
                .setTracerProvider(sdkTracerProvider)
                .setPropagators(contextPropagators)
                .build();
        OpenTelemetryAppender.install(openTelemetrySdk);
        return openTelemetrySdk;
    }
```
Can you please provide some information related to my query.

Thanks
Raghu Ram

**Reply by laurit** [reply]
> Also If i try to use the OtlpHttpLogExporter instead of OtlpHttpLogRecordExporter I am getting error that class was not found.

OtlpHttpLogExporter was renamed to OtlpHttpLogRecordExporter couple of years ago.

> Any idea why i can only see the metrics but not the original logs itself.

Please provide a minimal application that reproduces your issue along with any necessary instructions.

**Reply by yatamraghu7** [reply]
Hi @laurit 

Thanks a lot for replying to my messages all the time. I have recreated the problem with some sample project which is available here 
https://github.com/yatamraghu7/opentelemetry-poc/tree/main 
Can you please checkout my code and test if the logs that was being generated by my log4j is being passed to otel-collector or loki as I was not successful in doing this by myself.
I am using Java 17 version, Spring boot 2.7.1 and Apache Maven Version as 3.9.8
Kindly let me know if you need any further information.
I am using this link as reference to forward my logs to loki through open-telemtry collector
https://github.com/nlinhvu/opentelemetry-order-service-2023/blob/main/build.gradle

Thanks
Raghu

**Reply by yatamraghu7** [reply]
Hi @laurit 
My main target is to send the logs which log4j is writing in my spring boot application to otel-collector so that I can forward those logs to clickhouse database and the query them using qryn backend. 
Instead of jumping into the direct clickhouse i wanted to see if my logs are forwarded to otel-collector or not and would like to see the logs in grafana with the help of loki. unfortunately i was not able to forward the logs to loki.
Please help me in understanding this issue and find a fix for this.
Thanks

**Reply by laurit** [reply]
The simplest way to get this working is to either use the java agent https://opentelemetry.io/docs/zero-code/java/agent/ or otel spring starter https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/ It is possible to make this work without using these and doing things manually, but it does require some level of understanding of what is going on. I'd advise you to start with either the agent or spring starter to get started. To use either agent or starter start with deleting
```
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-starter-sleuth</artifactId>
			<version>3.1.3</version>
			<exclusions>
				<exclusion>
					<groupId>org.springframework.cloud</groupId>
					<artifactId>spring-cloud-sleuth-brave</artifactId>
				</exclusion>
			</exclusions>
		</dependency>
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-sleuth-otel-autoconfigure</artifactId>
			<version>1.1.2</version>
		</dependency>
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-sleuth-instrumentation</artifactId>
			<version>3.1.3</version>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry</groupId>
			<artifactId>opentelemetry-exporter-otlp</artifactId>
			<version>1.29.0</version>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry</groupId>
			<artifactId>opentelemetry-exporter-jaeger</artifactId>
			<version>1.29.0</version>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry.javaagent.instrumentation</groupId>
			<artifactId>opentelemetry-javaagent-opentelemetry-api-1.0</artifactId>
			<version>2.6.0-alpha</version>
			<scope>compile</scope>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry.javaagent</groupId>
			<artifactId>opentelemetry-javaagent</artifactId>
			<version>2.6.0</version>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-log4j-appender-2.17</artifactId>
			<version>2.6.0-alpha</version>
			<scope>compile</scope>
		</dependency> 
```
otel and sleuth both do tracing, you won't need both of them. Also delete `OpenTelemetryConfig` class, in your current setup I don't think it does anything. For spring starter follow https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/getting-started/ add
```
	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>io.opentelemetry.instrumentation</groupId>
				<artifactId>opentelemetry-instrumentation-bom</artifactId>
				<version>2.6.0</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>
```
and
```
		<dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-spring-boot-starter</artifactId>
		</dependency>
```
and you should be all set. For java agent instead of adding the dependencies you'd need to add the java agent to jvm arguments.
One thing to note is that although you are using log4j api to log the actual logging is done with logback not log4j so your log4j configuration doesn't do anything.

**Reply by yatamraghu7** [reply]
Hi @laurit 
I have followed how you recommended and i could not see the logs being forwarded to otel-collector. I did checkout my code just now can you please check if i am doing something wrong.
Also it would be great if you can check the docker file as well which I am using for otel-collector
Thanks
Raghu Ram

**Reply by laurit** [reply]
I did not try sending the logs to collector, I added `-Dotel.logs.exporter=console` to jvm arguments and saw the logs in stdout.

**Reply by yatamraghu7** [reply]
Hi @laurit 

Thanks for your help in this topic.
As I can see the logs are getting exported to my OTEL-COLLECTOR. Now i have few issues with the exporting.

I am trying to run my application as below
java -jar \
-Dotel.exporter.otlp.endpoint=http://localhost:4317 \
-Dotel.traces.exporter=otlp \
-Dotel.logs.exporter=otlp \
target/telemetry-poc-0.0.1-SNAPSHOT.jar \    
--spring.application.name=opentelemetry-poc \
--server.port=8099

I can see errors as follows in my terminal

2024-07-29 11:41:07.780  INFO [telemetry-poc,,] 3144 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 3307 ms
2024-07-29 11:41:08.410  WARN [telemetry-poc,,] 3144 --- [           main] i.o.i.r.ManifestResourceProvider         : Error reading manifest

java.nio.file.NoSuchFileException: -Dotel.exporter.otlp.endpoint=http:/localhost:4318
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92) ~[na:na]
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106) ~[na:na]
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[na:na]
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55) ~[na:na]
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:148) ~[na:na]
	at java.base/java.nio.file.Files.readAttributes(Files.java:1851) ~[na:na]
	at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1428) ~[na:na]
	at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:718) ~[na:na]
	at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:252) ~[na:na]
	at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:181) ~[na:na]
	at java.base/java.util.jar.JarFile.<init>(JarFile.java:346) ~[na:na]
	at java.base/java.util.jar.JarFile.<init>(JarFile.java:317) ~[na:na]
	at java.base/java.util.jar.JarFile.<init>(JarFile.java:297) ~[na:na]
	at io.opentelemetry.instrumentation.resources.ManifestResourceProvider.readManifest(ManifestResourceProvider.java:73) ~[opentelemetry-resources-2.6.0-alpha.jar!/:2.6.0-alpha]
	at java.base/java.util.Optional.flatMap(Optional.java:289) ~[na:na]


Also my traces are getting modified when I see the traces in my backend which is Qryn.
Exported logs are not that clear which is almost similar to metrics but not my real logs when I used the query builder in my qryn.
My format of exporting is 
Spring Application -> OTEL-COLLECTOR -> CLOUD HOUSE -> QRYN

Attached some of the issues screenshots.

Thanks
Raghu Ram

**Reply by laurit** [reply]
The -D flags should come before `-jar`. `-Dotel.traces.exporter=otlp` and `-Dotel.logs.exporter=otlp` are default values, you don't need to explicitly set them. `-Dotel.exporter.otlp.endpoint=http://localhost:4317/` 4317 is the otlp/grpc port. If you wish to use it you also need to set the otlp protocol to grpc. The default is http:/localhost:4318 with otlp/protobuf.

> Also my traces are getting modified when I see the traces in my backend which is Qryn.

you need to elaborate on this

> Exported logs are not that clear which is almost similar to metrics but not my real logs when I used the query builder in my qryn.

Please elaborate. I don't know anything about qryn.

> Attached some of the issues screenshots.

You forgot to attach the screenshot.

**Reply by yatamraghu7** [reply]
Hi @laurit 

I am sending curl as below with X-B3 header to pass my custom traceId and spanId and when I see the traces in my qryn backend the traceid is modified 

H 'Content-Type: application/json' \
  -H 'X-B3-TraceId: 1111111111111111' \
  -H 'X-B3-SpanId: 2222222222222222' \
  -H 'X-B3-Sampled: 1' \
  -d '{
  "id":7,
  "name":"raghu"
  }'

<img width="1435" alt="Screenshot 2024-07-29 at 12 46 22 PM" src="https://github.com/user-attachments/assets/ea12f3cf-e47e-4b9c-a2e2-81e2780a4416">

Can you please help me how can i get the traceId which I am sending with my request header into my backend trace viewer.

The logs which I was telling before for a specific service If I filter my query with application name I can only see the metrics.
Attached the screenshots.
<img width="1435" alt="Screenshot 2024-07-29 at 11 49 56 AM" src="https://github.com/user-attachments/assets/25a39dcc-ffb4-4473-84f5-8cb3c1c23de6">
<img width="1435" alt="Screenshot 2024-07-29 at 11 49 48 AM" src="https://github.com/user-attachments/assets/a41d43a2-14f5-40a4-a541-4b60f4f10686">
<img width="1435" alt="Screenshot 2024-07-29 at 11 49 27 AM" src="https://github.com/user-attachments/assets/84022bd8-9e7a-42eb-bb16-d22881f8e2ee">


Thanks
Raghu Ram

**Reply by laurit** [reply]
> I am sending curl as below with X-B3 header to pass my custom traceId and spanId and when I see the traces in my qryn backend the traceid is modified

By default otel uses w3c trace context headers to propagate context. I'd guess that the b3 hears that you use are just getting ignored. See https://opentelemetry.io/docs/languages/java/configuration/#propagators

**Reply by yatamraghu7** [reply]
Hi @laurit 

Thanks for the help and now I am successfully able to export the logs and traces from my application to otel-collector.
One last question would be do you have any idea how we can export the TraceId and spanId which I can see in my application at INFO and DEBUG level to otel-Collector.
I can only see the logs but not the traceId and spanId. It would be great if you have any document reference.

Thanks
Raghu Ram

**Reply by laurit** [reply, marked as answer]
Trace and span id are part of the otlp payload. Only when the log message is emitted within a trace will it have a trace/span id. You can confirm this by turning on verbose output in otel collector https://opentelemetry.io/docs/collector/troubleshooting/ If you don't see the trace/span id in your observability backend but see it in the collector then please seek assistance from appropriate channels for your observability backend.
